src/vs/workbench/contrib/terminal/browser/terminalService.ts

```diff
@@ -132,6 +137,38 @@ export class TerminalService implements ITerminalService {
 		this._handleInstanceContextKeys();
 		this._processSupportContextKey = KEYBINDING_CONTEXT_TERMINAL_PROCESS_SUPPORTED.bindTo(this._contextKeyService);
 		this._processSupportContextKey.set(!isWeb || this._remoteAgentService.getConnection() !== null);
+
+		const enableTerminalReconnection = this.configHelper.config.enablePersistentSessions;
+		const serverSpawn = this.configHelper.config.serverSpawn;
+		if (!!this._environmentService.remoteAuthority && enableTerminalReconnection && serverSpawn) {
+			this._remoteTerminalsInitialized = this._reconnectToRemoteTerminals();
+			this._connectionState = TerminalConnectionState.Connecting;
+		} else {
+			this._connectionState = TerminalConnectionState.Connected;
+		}
+	}
+
+	private async _reconnectToRemoteTerminals(): Promise<void> {
+		const remoteTerms = await this._remoteTerminalService.listTerminals();
+		this._connectionState = TerminalConnectionState.Connected;
+		this._onDidChangeConnectionState.fire();
```

Comment:
Should this go at the end of the method after all the terminals have been created? Not sure how this is working atm because I thought event handlers all trigger synchronously on fire?

Reply:
They do, but I think we don't actually need the event. We should set the state first, then we will add a new terminal, `onInstancesChanged` will fire, and that will cause the title to update.